32blit-stm32/Src/SystemMenu/system_menu_controller.cpp

```diff
@@ -0,0 +1,136 @@
+/* SystemMenuController.cpp
+ * source file for Firmware menu
+ *
+ * When the user presses the MENU button, the system shows a menu on top of the content. This file
+ * contains functions to help drawing this menu. The actual invoking code is still in 32blit.cpp
+ * because that code has too many dependencies on the firmware code.
+ */
+
+#include "32blit.h"
+#include "32blit.hpp"
+#include <algorithm>
+
+using namespace blit;
+
+#include "system_menu_controller.hpp"
+#include "firmware_menu.hpp"
+#include "battery_menu.hpp"
+#include "engine/version.hpp"
+
+//
+// Initialize the system menu controller
+//
+SystemMenuController::SystemMenuController() {
+  current_menu = SystemMenus::Firmware;
+}
+
+//
+// Preapre to show the system menus
+//
+void SystemMenuController::prepare() {
+  set_menu(SystemMenus::Firmware);
+
+  firmware_menu.prepare();
+  battery_menu.prepare();
+}
+
+//
+// Call update on the currently selected menu
+//
+void SystemMenuController::update(uint32_t time) {
+  switch (current_menu) {
+  case SystemMenus::Firmware:
+    firmware_menu.update(time);
+    break;
+  case SystemMenus::Battery:
+    battery_menu.update(time);
+    break;
+  }
+}
+
+//
+// Render the currently selected menu
+//
+void SystemMenuController::render(uint32_t time) {
+  switch (current_menu) {
+  case SystemMenus::Firmware:
+    firmware_menu.render();
+    break;
+  case SystemMenus::Battery:
+    battery_menu.render();
+    break;
+  }
+
+  render_header_battery_status(time);
+  render_footer_version(time);
+}
+
+//
+// Render the battery bar
+//
+void SystemMenuController::render_header_battery_status(uint32_t time) {
+  const int screen_width = blit::screen.bounds.w;
+  const int screen_height = blit::screen.bounds.h;
+
+  //   const Pen foreground_colour = get_menu_colour(10);
+  const Pen bar_background_color = get_menu_colour(3);
+
+  BatteryInformation bat = blit_get_battery_info();
+
+  screen.text("bat", minimal_font, Point(screen_width - 80, 4));
```

Comment:
This text isn't visible any more. Wrong pen?

Reply:
Yes, most likely. When I moved code around, the pens would get set to the background and not reset to foreground. I fixed multiple similar  pen errors, but did not catch this one.